mesh/MeshNormals.cs

```diff
@@ -59,51 +64,48 @@ public void CopyTo(DMesh3 SetMesh)
-                Vector3d vc = Mesh.GetVertex(tri.c);
-                Vector3d N = MathUtil.Normal(ref va, ref vb, ref vc);
-                double a = MathUtil.Area(ref va, ref vb, ref vc);
-                bool taken = false;
-                Normals_lock.Enter(ref taken);
-                Normals[tri.a] += a * N;
-                Normals[tri.b] += a * N;
-                Normals[tri.c] += a * N;
-                Normals_lock.Exit();
-            }, maxDegreeOfParallelism);
-
-            gParallel.BlockStartEnd(0, NV - 1, (vi_start, vi_end) => {
-                for (int vi = vi_start; vi <= vi_end; vi++) {
-                    if (Normals[vi].LengthSquared > MathUtil.ZeroTolerancef)
-                        Normals[vi] = Normals[vi].Normalized;
-                }
-            }, maxDegreeOfParallelism);
+            arrayPool ??= ArrayPool<Vector3f>.Shared;
+            Vector3f[] normalsArray = arrayPool.Rent(mesh.MaxVertexID);
+            QuickComputeToArray(mesh, normalsArray);
+            // Write to mesh
+            mesh.EnableVertexNormals(Vector3f.Zero);
+            for (int vid = 0; vid < mesh.MaxVertexID; vid++)
+            {
+                if (!mesh.IsVertex(vid))
+                    continue;
+                mesh.SetVertexNormal(vid, normalsArray[vid]);
+            }
+            arrayPool.Return(normalsArray);
         }
 
-
-
-
-        public static void QuickCompute(DMesh3 mesh, int maxDegreeOfParallelism)
+        private static void QuickComputeToArray(DMesh3 mesh, Vector3f[] normalsArray)
         {
-            MeshNormals normals = new MeshNormals(mesh);
-            normals.Compute(maxDegreeOfParallelism);
-            normals.CopyTo(mesh);
+            for (int vid = 0; vid < mesh.MaxVertexID; vid++)
+                normalsArray[vid] = Vector3f.Zero;
+            for (int tid = 0; tid < mesh.MaxTriangleID; tid++)
+            {
+                if (!mesh.IsTriangle(tid))
+                    continue;
+                Index3i triangle = mesh.GetTriangle(tid);
+                Vector3d vertexA = mesh.GetVertexUnsafe(triangle.a);
+                Vector3d vertexB = mesh.GetVertexUnsafe(triangle.b);
+                Vector3d vertexC = mesh.GetVertexUnsafe(triangle.c);
+                Vector3d triangleNormal = MathUtil.Normal(vertexA, vertexB, vertexC);
+                double triangleArea = MathUtil.Area(vertexA, vertexB, vertexC);
+                Vector3f areaNormalMultiplication = (Vector3f)(triangleArea * triangleNormal);
+                normalsArray[triangle.a] += areaNormalMultiplication;
+                normalsArray[triangle.b] += areaNormalMultiplication;
+                normalsArray[triangle.c] += areaNormalMultiplication;
+            }
+            // Normalize array
+            for (int vid = 0; vid < mesh.MaxVertexID; vid++)
+            {
+                Vector3f vertexNormal = normalsArray[vid];
+                if (vertexNormal.LengthSquared > MathUtil.ZeroTolerancef)
```

Comment:
instead of LengthSquared you can use LengthL1, which may work faster

Reply:
The `vertexNormal != Vector3f.Zero` is even simpler and covers the condition correctly